compiler/rustc_mir/src/transform/instrument_coverage.rs

```diff
@@ -303,86 +519,266 @@ impl<'a, 'tcx> Instrumentor<'a, 'tcx> {
+                        next iter, and add prev={:?}",
+                        prev
+                    );
+                    pending_dup_spans.push(prev);
+                }
+            } else {
+                debug!(
+                    "  different bcbs, overlapping spans, so ignore/drop pending and only add prev \
+                    if it has statements that end before curr={:?}",
+                    prev
+                );
+                pending_dup_spans.clear();
+                prev.cutoff_statements_at(curr.span.lo());
+                if !prev.coverage_statements.is_empty() {
+                    debug!("  ... adding modified prev={:?}", prev);
+                    coverage_spans.push(prev);
+                } else {
+                    debug!("  ... no non-overlapping statements to add");
+                }
+            }
+
+            prev = curr;
+        }
+        debug!("    AT END, adding last prev={:?}", prev);
+        for pending in pending_dup_spans.drain(..) {
+            debug!("    ...adding at least one pending={:?}", pending);
+            coverage_spans.push(pending);
+        }
+        coverage_spans.push(prev);
+
+        // FIXME(richkadel): Replace some counters with expressions if they can be calculated based
+        // on branching. (For example, one branch of a SwitchInt can be computed from the counter
+        // for the CoverageSpan just prior to the SwitchInt minus the sum of the counters of all
+        // other branches).
+        coverage_spans
+    }
+
+    /// Converts the computed `BasicCoverageBlock`s into `SpanViewable`s.
+    fn span_viewables(&self, coverage_spans: &Vec<CoverageSpan>) -> Vec<SpanViewable> {
+        let tcx = self.tcx;
+        let mir_body = &self.mir_body;
+        let mut span_viewables = Vec::new();
+        for coverage_span in coverage_spans {
+            let bcb = self.bcb_from_coverage_span(coverage_span);
+            let CoverageSpan { span, bcb_leader_bb: bb, coverage_statements } = coverage_span;
+            let id = bcb.id();
+            let tooltip = coverage_statements
+                .iter()
+                .map(|covstmt| covstmt.format(tcx, mir_body))
+                .collect::<Vec<_>>()
+                .join("\n");
+            span_viewables.push(SpanViewable { bb: *bb, span: *span, id, tooltip });
+        }
+        span_viewables
+    }
+
+    #[inline(always)]
+    fn bcb_from_coverage_span(&self, coverage_span: &CoverageSpan) -> &BasicCoverageBlock {
+        &self.basic_coverage_blocks()[coverage_span.bcb_leader_bb]
+    }
+
+    #[inline(always)]
+    fn body_span(&self) -> Span {
+        self.hir_body.value.span
+    }
+
+    // Return the de-duplicated spans of a block or blocks. Multiple blocks represent a segment of
+    // the MIR control-flow-graph that have no branches. For example: A; A -> B; A -> B -> C; and
+    // and so on.
```

Comment:
I think this is the definition of BasicCoverageBlock?

Reply:
Great catch. Yep. The comment is outdated. I'm updating it.

Reply:
Comments updated